src/lib/IDBDAO.js

```diff
@@ -0,0 +1,350 @@
+          };
+        });
+      }
+    },
+  ],
+
+  methods: [
+
+    function init() {
+
+      var propKeys = this.propKeys_ = {};
+      var properties = this.of.getAxiomsByClass(foam.core.Property);
+      for ( var i = 0 ; i < properties.length ; i++ ) {
+        var prop = properties[i];
+        propKeys[prop.name] = prop.name;
+        if ( prop.shortName ) propKeys[prop.shortName] = prop.name;
+      }
+    },
+
+    function deserialize(json) {
+      return foam.json.parse(foam.json.parseString(json)); //TODO: serialization
+    },
+
+    function serialize(obj) {
+      return foam.json.stringify(obj); //TODO: serialization
+    },
+
+    function withStore(mode, fn) {
+      if ( mode !== 'readwrite' ) return this.withStore_(mode, fn);
+
+      var self = this;
+
+      if ( ! this.q_ ) {
+        var q = [fn];
+        this.q_ = q;
+        this.async(function() {
+          self.withStore_(mode, function(store) {
+            // console.log('q length: ', q.length);
+            if ( self.q_ == q ) self.q_ = undefined;
+            for ( var i = 0 ; i < q.length ; i++ ) q[i](store);
+          });
+        })();
+      } else {
+        this.q_.push(fn);
+        // Diminishing returns after 10000 per batch
+        if ( this.q_.length == 10000 ) this.q_ = undefined;
+      }
+    },
+
+    function withStore_(mode, fn) {
+      var self = this;
+      if ( self.__TXN__[0] ) {
+        try {
+          fn.call(self, self.__TXN__[0]);
+          return;
+        } catch (x) {
+          self.__TXN__[0] = undefined;
+        }
+      }
+      self.withDB.then(function (db) {
+        var tx = db.transaction([self.name], mode);
+        var os = tx.objectStore(self.name);
+        self.__TXN__[0] = os;
+        fn.call(self, os);
+      });
+    },
+
+    function put(value) {
+      var self = this;
+      return new Promise(function(resolve, reject) {
+        self.withStore("readwrite", function(store) {
+          var request = store.put(self.serialize(value), value.id);
+          request.transaction.addEventListener(
+            'complete',
+            function(e) {
+              self.pub('on','put', value);
+              resolve(value);
+            });
+          request.transaction.addEventListener(
+            'error',
+            function(e) {
+              reject(self.IDBInternalException.create({ id: value.id, error: e }));
+            });
+        });
+      });
+    },
+
+    function find(key) {
+      var self = this;
+
+      return new Promise(function(resolve, reject) {
+        self.withStore("readonly", function(store) {
+          var request = store.get(key);
+          request.transaction.addEventListener(
+            'complete',
+            function() {
+              if (!request.result) {
+                reject(self.ObjectNotFoundException.create({ id: key }));
+                return;
+              }
+              var result = self.deserialize(request.result);
```

Comment:
Let's put a todo somewhere that we should treat catch and handle errors from deserialization/serialization as internal errors and surface them to the users as such.